### PR TITLE
Reland (#120419):[Exegesis][RISCV] Add RISCV support for llvm-exegesis

### PR DESCRIPTION
**Pull request by dybv-sc**
2nd reland attempt. This should be merged again and applying https://github.com/llvm/llvm-project/pull/120451 should fix buildbot issue.

Original commit message:
```
[Exegesis][RISCV] Add RISCV support for llvm-exegesis (https://github.com/llvm/llvm-project/pull/89047)
This patch also makes following amendments to core exegesis:
* Added distinction between regular registers aliasing check and
registers used as memory address in instruction.
* Added scratch memory space pointer register.
* General exegesis options were amended:
        * mattr - new option to pass a list of enabled target features

Llvm-exegesis RISCV port is a result of team effort. Below everyone
involved listed.
Co-authored-by: Konstantin Vladimirov
<konstantin.vladimirov@syntacore.com>
Co-authored-by: Dmitrii Petrov <dmitrii.petrov@syntacore.com>
Co-authored-by: Dmitry Bushev <dmitry.bushev@syntacore.com>
Co-authored-by: Mark Goncharov <mark.goncharov@syntacore.com>
Co-authored-by: Anastasiya Chernikova
<anastasiya.chernikova@syntacore.com>

---------

Co-authored-by: Dmitry Bushev <dmitry.bushev@syntacore.com>
```

### CODE REVIEW

**Comment by dybv-sc** [comment]
@boomanaiden154 , @thurstond, @kazutakahirata FYI

**Comment by dybv-sc** [comment]
> @dybv-sc Thanks for the update! When I try out your patch, I am getting:
> 
> ```
> CMake Error at cmake/modules/LLVMProcessSources.cmake:113 (message):
>   Found erroneous configuration for source file RISCVCounters.cpp
> 
>   LLVM's build system enforces that all source files are added to a build
>   target, that exactly one build target exists in each directory, and that
>   this target lists all files in that directory.  If you want multiple
>   targets in the same directory, add PARTIAL_SOURCES_INTENDED to the target
>   specification, though it is discouraged.
> ```
> 
> I'm guessing that `llvm/tools/llvm-exegesis/lib/RISCV/CMakeLists.txt` should somehow mention `RISCVCounters.cpp`. Thoughts?

Hi,

Have you applied all of the patches there are in this request? We used to have this file, but after some discussion, we removed it (with the intention to rework and bring it back in some later request). Currently there should be no this file on top of our branch.

File: llvm/tools/llvm-exegesis/lib/RISCV/Target.cpp [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+  generateInstructionVariants(const Instruction &Instr,
+                              unsigned MaxConfigsPerOpcode) const override;
+};
+
+ExegesisRISCVTarget::ExegesisRISCVTarget()
+    : ExegesisTarget(ArrayRef<CpuAndPfmCounters>{},
+                     RISCV_MC::isOpcodeAvailable) {}
+
+#define GET_REGISTER_MATCHER
+#include "RISCVGenAsmMatcher.inc"
```
Comment (boomanaiden154) [inline llvm/tools/llvm-exegesis/lib/RISCV/Target.cpp:139]: This should be removed before relanding. It doesn't make sense to land with a forward fix already planned.

Comment (dybv-sc) [reply llvm/tools/llvm-exegesis/lib/RISCV/Target.cpp]: You are right, amended Kazu's fix

**Review by boomanaiden154 — APPROVED** [review]
LGTM.

File: llvm/tools/llvm-exegesis/lib/RISCV/Target.cpp (hunk 1 above)
Comment (thurstond) [reply llvm/tools/llvm-exegesis/lib/RISCV/Target.cpp]: LGTM. Please update the commit message before merging.

**Review by thurstond — APPROVED** [review]

**Comment by kazutakahirata** [comment]
> > @dybv-sc Thanks for the update! When I try out your patch, I am getting:
> > ```
> > CMake Error at cmake/modules/LLVMProcessSources.cmake:113 (message):
> >   Found erroneous configuration for source file RISCVCounters.cpp
> > 
> >   LLVM's build system enforces that all source files are added to a build
> >   target, that exactly one build target exists in each directory, and that
> >   this target lists all files in that directory.  If you want multiple
> >   targets in the same directory, add PARTIAL_SOURCES_INTENDED to the target
> >   specification, though it is discouraged.
> > ```
> >   
> > I'm guessing that `llvm/tools/llvm-exegesis/lib/RISCV/CMakeLists.txt` should somehow mention `RISCVCounters.cpp`. Thoughts?
> 
> Hi,
> 
> Have you applied all of the patches there are in this request? We used to have this file, but after some discussion, we removed it (with the intention to rework and bring it back in some later request). Currently there should be no this file on top of our branch.

Huh.  When I applied `https://github.com/llvm/llvm-project/pull/120467.patch` locally, that didn't work.  When I applied `https://github.com/llvm/llvm-project/pull/120467.diff`, however, that did work.  I didn't know there was a difference between `patch` and `diff`.

Anyway, thank you for updating the patch!

**Review by kazutakahirata — APPROVED** [review]
LGTM.  Thanks!

**Pull request merged**